src/common/io.ts

```diff
@@ -1716,6 +1718,108 @@ export class ARIOWriteable extends ARIOReadable implements AoARIOWrite {
     });
   }
 
+  async setPrimaryName(
+    params: AoArNSPurchaseParams,
+    options?: WriteOptions<
+      keyof SetPrimaryNameProgressEvents,
+      SetPrimaryNameProgressEvents[keyof SetPrimaryNameProgressEvents]
+    >,
+  ): Promise<AoMessageResult> {
+    options?.onSigningProgress?.('requesting-primary-name', {
+      name: params.name,
+      fundFrom: params.fundFrom,
+      referrer: params.referrer,
+    });
+
+    // create the primary name request, if it already exists, get the request and base name owner
```

Comment:
TODO: support `fundFrom` as `turbo` here to create the primary name request - but i am not certain the error path if the request already exists. will need to test more.